apps/ledger-live-desktop/src/renderer/components/CustomImage/ImageGrayscalePreview.tsx

```diff
@@ -0,0 +1,219 @@
+      /**
+       * Floyd-Steinberg dithering
+       * https://en.wikipedia.org/wiki/Floyd%E2%80%93Steinberg_dithering
+       *         x - 1  |   x    |  x + 1
+       *   y            |   *    | 7 / 16
+       * y + 1  3 / 16  | 5 / 16 | 1 / 16
+       */
+      const newpixel = posterizedGray256;
+      pixels256[y][x] = newpixel;
+      if (dither) {
+        const quantError = oldpixel - newpixel;
+        if (x < width - 1) {
+          pixels256[y][x + 1] = Math.floor(pixels256[y][x + 1] + quantError * (7 / 16));
+        }
+        if (x > 0 && y < height - 1) {
+          pixels256[y + 1][x - 1] = Math.floor(pixels256[y + 1][x - 1] + quantError * (3 / 16));
+        }
+        if (y < height - 1) {
+          pixels256[y + 1][x] = Math.floor(pixels256[y + 1][x] + quantError * (5 / 16));
+        }
+        if (x < width - 1 && y < height - 1) {
+          pixels256[y + 1][x + 1] = Math.floor(pixels256[y + 1][x + 1] + quantError * (1 / 16));
+        }
+      }
+
+      const val16 = clamp(Math.floor(pixels256[y][x] / rgbStep), 0, 16 - 1);
+      pixels16[y][x] = val16;
+      /** gray rgb value after applying the contrast, in [0,255] */
+      const val256 = val16 * rgbStep;
+      filteredImageData.push(val256); // R
+      filteredImageData.push(val256); // G
+      filteredImageData.push(val256); // B
+      filteredImageData.push(255); // alpha
+    }
+  }
+
+  // Raw data -> by column, from right to left, from top to bottom
+  for (let x = width - 1; x >= 0; x--) {
+    for (let y = 0; y < height; y++) {
+      const val16 = pixels16[y][x];
+      const grayHex = val16.toString(16);
+      hexRawResult = hexRawResult.concat(grayHex);
+    }
+  }
+
+  return {
+    imageDataResult: Uint8ClampedArray.from(filteredImageData),
+    hexRawResult,
+  };
+}
+
+function processImage(args: { image: HTMLImageElement; contrast: number }): ProcessorResult {
+  const { image, contrast } = args;
+  const { context, canvas } = createCanvas(image);
+  if (!context) throw Error("Context is null");
+  context.drawImage(image, 0, 0);
+  const imageData = context.getImageData(0, 0, canvas.width, canvas.height);
+  const { naturalHeight: height, naturalWidth: width } = image;
+  // 2. applying filter to the image data
+  const { imageDataResult: grayData, hexRawResult } = applyFilter(imageData, contrast);
+  context.putImageData(
+    new ImageData(grayData, width, height), // eslint-disable-line no-undef
+    0,
+    0,
+  );
+  const grayScaleBase64 = canvas.toDataURL();
+  return {
+    previewResult: { imageBase64DataUri: grayScaleBase64, height, width },
+    rawResult: { hexData: hexRawResult, height, width },
+  };
+}
+
+export type Props = ImageBase64Data & {
+  onError: (_: Error) => void;
+  onResult: (_: ProcessorResult) => void;
+  setLoading: (_: boolean) => void;
+  /**
+   * number >= 0
+   *  - 0:  full black
+   *  - 1:  original contrast
+   *  - >1: more contrasted than the original
+   * */
```

Comment:
is this comment on `number` still relevant ?

Reply:
it's misplaced, thanks!